rainier-core/src/main/scala/com/stripe/rainier/compute/Target.scala

```diff
@@ -47,6 +47,29 @@ class Target(val real: Real, val placeholders: Map[Variable, Array[Double]]) {
       Some(result)
     }
 
+  def batched(batchBits: Int): (List[Variable], List[Real]) = {
+    val (variables, outputs) =
+      0.to(batchBits)
+        .toList
+        .map { i =>
+          batch(i)
+        }
+        .unzip
+
+    (placeholderVariables ++ variables.flatten, real :: outputs)
+  }
+
+  private def batch(bit: Int): (List[Variable], Real) =
+    0.until(1 << bit).foldLeft((List.empty[Variable], Real.zero)) {
+      case ((v, r), _) =>
+        val newVars = placeholderVariables.map { _ =>
+          new Variable
+        }
+        val newReal =
+          PartialEvaluator.inline(real, placeholderVariables.zip(newVars).toMap)
```

Comment:
In the PR comment you mention that each copy of the function is restricted to computing a 2^k observations but I don't see how this is implemented. Do I see it right that a single batch creates `2^bit` copies of exactly the same function?

If so:

1. Wouldn't it be more direct to say?
```
List.fill(2^bit) {
  val newVars = ...
  ...
}
```

2. is there any concern over code size blowup? or is packing of IR supposed to take care of it by reusing most of the IR?

Reply:
Yes, it's 2^bit copies whose results are then summed.
1. yep, that would be cleaner, thanks
2. I am indeed a bit worried about code size blowup. Depending on the exact function it may be able to optimize or reuse large parts of it, but there's no guarantee of that. In my quick empirical testing, you get most of the benefit from just 2^4 copies, which doesn't seem *that* bad.

Reply:
If a small number of copies is enough, does it make sense to cap them?